### PR TITLE
Fix bugs in scale_batch_size

### LINKED ISSUE #2484: Trainer.scale_batch_size requires model.batch_size instead of model.hparams.batch_size
## 🐛 Bug

`Trainer.scale_batch_size` only works if a model has the `batch_size` property and does not work with `model.hparams.batch_size` even though all documentation points to the reverse.

### To Reproduce

All of my hyperparameters are available as model.hparams like suggested in the documentation: ([hyperparameters, option 3](https://pytorch-lightning.readthedocs.io/en/0.8.4/hyperparameters.html#lightningmodule-hyperparameters).
This means that my `batch_size` is available as `model.hparams.batch_size`.

This should be fully compatible with the [documented example code](https://pytorch-lightning.readthedocs.io/en/0.8.4/training_tricks.html#auto-scaling-of-batch-size) of `Trainer.scale_batch_size()` since that code also uses `model.hparams.batch_size` instead of `model.batch_size`.

However, when I put my model in `Trainer.scale_batch_size`, I get the following error:

    pytorch_lightning.utilities.exceptions.MisconfigurationException: Field batch_size not found in `model.hparams`

### Example code

```
class LitModel(pl.LightningModule):
    def __init__(self, hparams):
        super().__init__()
        self.hparams = args

model = LitModel(args)
trainer = Trainer()
trainer.scale_batch_size(model)
```

### Expected behavior

Either `Trainer.scale_batch_size` should work with `model.hparams` or the error message, linked documentation examples and docstrings should all change (i.e. [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_tricks.py#L139), [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_tricks.py#L228) and [here](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/training_tricks.py#L233)).

(I would prefer the second option. I think that it should work with both `model.batch_size` and `model.hparams.batch_size`.)

### Environment

 - pytorch-lightning         0.8.4

### PR DESCRIPTION
**Pull request by x2-l**
Fixes #2484
- Set model.batch_size based on model.hparams.batch_size if not defined
- Changed init_val for batch_size to 0 so that it starts with user defined batch_size instead of 2 all the time

### CODE REVIEW

**Comment by SkafteNicki** [comment]
I am not sure this will solve your problem. Since it is only `model.batch_size` that gets updated during the scaling, if you have defined your dataloaders using `Dataloader(dataset, batch_size=self.hparams.batch_size)` then this will have no effect.

**Comment by x2-l** [comment]
this should be handled by user no? its the same as enabling `auto_lr_find=True` but the user sets `Optimizer(self.parameters(), lr=self.hparams.learning_rate)`. or maybe `scale_batch_size` should consider overwritting `DataLoader.batch_size` if it is called?

**Comment by justusschock** [comment]
Maybe we also need to check, if self.hparams is a mapping, because then you can't use `setattr`

**Comment by SkafteNicki** [comment]
@x2-l `DataLoader.batch_size` cannot be overwritten after it is initialized, already tried it because that would be the easiest solution.
@justusschock agree, in PR #1998 I have implemented three functions as counterparts to `setattr`, `hasattr` and `getattr` that basically search for attribute in this order:
1. attribute of `model` i.e. `model.batch_size`
2. key of `model.hparams` i.e. `model.hparams['batch_size']
3. attribute of `model.hparams` i.e. `model.hparams.batch_size`

**Comment by x2-l** [comment]
@SkafteNicki what bout reinitializing `DataLoader` with the new `batch_size` like 
```
kwargs = get_all_existing attribute from DataLoader
kwargs['batch_size] = searched_batch_size
dataloader = DataLoader(dataloader.dataset, **kwargs)
```

**Comment by SkafteNicki** [comment]
@x2-l agree that should work

**Comment by awaelchli** [comment]
@x2-l I think there is already code like this for replacing the sampler. One could factor this out and create a function
`replace_dataloader_attr(dataloader: DataLoader, batch_size: str)` like you proposed and avoid duplicated code.

**Comment by SkafteNicki** [comment]
Here is the function we already have: https://github.com/PyTorchLightning/pytorch-lightning/blob/927f305f7e556828b5cdd45e3977c67f3c54b8fc/pytorch_lightning/trainer/data_loading.py#L169-L178

**Comment by Borda** [comment]
I agree that this is quite a monkey patch and it would be better to support `'hparams.batch_size'` as `batch_arg_name`

**Comment by x2-l** [comment]
@SkafteNicki @Borda should I still update this? or can be updated together at PR #1998?

**Comment by x2-l** [comment]
on second thought, why not just raise an warning that `model.batch_size` will be used when both `model.batch_size` and `model.hparams.batch_size` are set? if only `model.hparams.batch_size` is set, the attribute `batch_size` will be created for `model`.

the `scale_batch_size` function will then update both params so that it'll be consistent no matter what is being used in the rest of the code. same can be done for the `learning_rate`.

@Borda @SkafteNicki @awaelchli thoughts?

**Comment by justusschock** [comment]
While I see your point, we should be careful about that. E.g. when you reinstantiate your model and overwrite the wrong parameter, you introduce unwanted behaviour somewhere else.

I think I'd prefer not to implicitly introduce concurring new arguments of not necessary

**Comment by Borda** [comment]
I would say, do this patch/fix here because it may take some time before PR #1998 lands...
also, this may be 0.8.x release compare to PR #1998 is aimed for 0.9

**Comment by x2-l** [comment]
I enabled scaling of batch size using `model.hparams. batch_arg_name` and removed setting the `. batch_arg_name` attribute on the model. If both are specified, a warning is raised telling user that `model. batch_arg_name` will be used as initial value for the search. If it's undesired, user can stop and change accordingly.

**Comment by Borda** [comment]
seems taking parts from #2223 so let's merge this after it...

**Comment by mdgoldberg** [comment]
Hello! Will this and/or #2223 also fix `auto_lr_find`? It has the same issue.

**Comment by Borda** [comment]
@ghost mind finish this one as #2223 have been merged...

**Comment by awaelchli** [comment]
Hey! recently a similar PR/issue was fixed by @SkafteNicki in #2821. Could we use the lightning_attr api he added there to implement these checks here? They look very similar. 
And we need to keep in mind the compatibility with datamodules, but perhaps this should be done in a different PR :)

**Comment by SkafteNicki** [comment]
Agree with @awaelchli that the first issue addressed in this PR is the same issue the learning rate finder had. If you could exchange all `hasattr`, `getattr`, `setattr` with the corresponding function from this file https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/utilities/parsing.py#L145-L195 then it would be great :)

**Comment by awaelchli** [comment]
It looks like this user and their repository completely vanished from github. Should we copy these changes over into a new PR and continue as we discussed?

**Comment by Borda** [comment]
> It looks like this user and their repository completely vanished from github. Should we copy these changes over into a new PR and continue as we discussed?

he behaves as his nick says :D
I would say, check if we can push/rebase this PR, if yes, let's continue if not let's make the same branch in our repo, and merge this PR into the new branch and continue there...
